File: myFinance_improved_V2/lib/features/cash_transaction/data/models/cash_location_model.dart
import '../../domain/entities/cash_location.dart';

/// Data model for CashLocation
///
/// Handles JSON serialization and mapping to domain entity
class CashLocationModel {
  final String cashLocationId;
  final String locationName;
  final String locationType;
  final String? storeId;
  final String? companyId;
  final String? accountId;

  CashLocationModel({
    required this.cashLocationId,
    required this.locationName,
    required this.locationType,
    this.storeId,
    this.companyId,
    this.accountId,
  });

  /// From JSON (API/Database) → Model
  /// RPC returns camelCase fields (id, name, type, storeId)
  /// Also supports snake_case for direct table queries
  factory CashLocationModel.fromJson(Map<String, dynamic> json) {
    // Get additionalData which contains snake_case fields from raw table
    final additionalData = json['additionalData'] as Map<String, dynamic>?;

    return CashLocationModel(
      // RPC returns 'id', table returns 'cash_location_id'
      cashLocationId: (json['id'] ?? json['cash_location_id'] ?? '') as String,
      // RPC returns 'name', table returns 'location_name'
      locationName: (json['name'] ?? json['location_name'] ?? '') as String,
      // RPC returns 'type', table returns 'location_type'
      locationType: (json['type'] ?? json['location_type'] ?? '') as String,
      // RPC returns 'storeId', table/additionalData returns 'store_id'
      storeId: (json['storeId'] ?? additionalData?['store_id']) as String?,
      // RPC returns company info in additionalData
      companyId: (json['companyId'] ?? additionalData?['company_id']) as String?,
      accountId: (json['accountId'] ?? json['account_id']) as String?,
    );
  }

  /// Model → Domain Entity
  CashLocation toEntity() {
    return CashLocation(
      cashLocationId: cashLocationId,
      locationName: locationName,
      locationType: locationType,
      storeId: storeId,
      companyId: companyId,
      accountId: accountId,
    );
  }

  /// Model → JSON (for API calls)
  Map<String, dynamic> toJson() {
    return {
      'cash_location_id': cashLocationId,
      'location_name': locationName,
      'location_type': locationType,
      'store_id': storeId,
      'company_id': companyId,
      'account_id': accountId,
    };
  }
}

File: website/src/features/cash_ending/data/models/CashLocationModel.ts
/**
 * CashLocationModel
 * Handles JSON transformation between API and Domain Entity
 */

import { CashLocationEntity, type CashLocation } from '../../domain/entities/CashLocation';

export class CashLocationModel {
  /**
   * Convert API JSON response to Domain Entity
   */
  static fromJson(json: CashLocation): CashLocationEntity {
    return new CashLocationEntity(
      json.cash_location_id,
      json.location_name,
      json.location_type,
      json.store_id,
      json.store_name,
      json.company_id,
      json.is_active,
      json.created_at,
      json.updated_at
    );
  }

  /**
   * Convert Domain Entity to API JSON format
   */
  static toJson(entity: CashLocationEntity): CashLocation {
    return {
      cash_location_id: entity.cashLocationId,
      location_name: entity.locationName,
      location_type: entity.locationType,
      store_id: entity.storeId,
      store_name: entity.storeName,
      company_id: entity.companyId,
      is_active: entity.isActive,
      created_at: entity.createdAt,
      updated_at: entity.updatedAt,
    };
  }
}
